File: app/src/renderer/src/lib/api/orchestration.ts
import { apiClient } from "./client";

// ── Types ────────────────────────────────────────────────────

export type OrchestrationSessionStatus = "pending" | "running" | "paused" | "completed" | "failed" | "cancelled";
export type OrchestrationTaskStatus = "pending" | "queued" | "running" | "completed" | "failed" | "skipped" | "cancelled";
export type ApprovalStatus = "pending" | "approved" | "rejected";

export type OrchestrationSessionRecord = {
  id: string;
  conversationId: string;
  mode: string;
  status: OrchestrationSessionStatus;
  sharedContext: Record<string, any> | null;
  createdBy: string | null;
  errorMessage: string | null;
  startedAt: string | null;
  completedAt: string | null;
  createdAt: string | null;
};

export type OrchestrationSessionDetail = OrchestrationSessionRecord & {
  tasks: OrchestrationTaskRecord[];
  approvals: OrchestrationApprovalRecord[];
};

export type OrchestrationTaskRecord = {
  id: string;
  workerRole: string;
  title: string;
  description: string | null;
  status: OrchestrationTaskStatus;
  dependsOn: string[];
  sequenceOrder: number;
  errorMessage: string | null;
  startedAt: string | null;
  completedAt: string | null;
};

export type OrchestrationApprovalRecord = {
  id: string;
  taskId: string;
  status: ApprovalStatus;
  reason: string | null;
  reviewerNotes: string | null;
  requestedAt: string | null;
  resolvedAt: string | null;
};

export type CheckpointRecord = {
  id: string;
  taskId: string;
  state: Record<string, any>;
  createdAt: string;
};

export type CreateSessionPayload = {
  conversation_id: string;
  mode?: string;
};

export type CreateTaskPayload = {
  worker_role: string;
  title: string;
  description?: string;
  input_context?: Record<string, any>;
  depends_on?: string[];
};

export type ApprovalResolvePayload = {
  approved: boolean;
  notes?: string;
};

// ── API ──────────────────────────────────────────────────────

export const orchestrationApi = {
  async createSession(payload: CreateSessionPayload): Promise<{ id: string; conversationId: string; mode: string; status: string; sharedContext: Record<string, any> | null; createdAt: string | null }> {
    return apiClient.post("/orchestration/sessions", payload);
  },

  async listSessions(params?: { conversation_id?: string; status?: string }): Promise<OrchestrationSessionRecord[]> {
    const query = new URLSearchParams();
    if (params?.conversation_id) query.append("conversation_id", params.conversation_id);
    if (params?.status) query.append("status", params.status);
    const qs = query.toString() ? `?${query.toString()}` : "";
    return apiClient.get<OrchestrationSessionRecord[]>(`/orchestration/sessions${qs}`);
  },

  async getSession(sessionId: string): Promise<OrchestrationSessionDetail> {
    return apiClient.get<OrchestrationSessionDetail>(`/orchestration/sessions/${sessionId}`);
  },

  async addTask(sessionId: string, payload: CreateTaskPayload): Promise<{ id: string; workerRole: string; title: string; status: string; sequenceOrder: number }> {
    return apiClient.post(`/orchestration/sessions/${sessionId}/tasks`, payload);
  },

  async executeSession(sessionId: string): Promise<{ id: string; status: string }> {
    return apiClient.post(`/orchestration/sessions/${sessionId}/execute`);
  },

  async cancelSession(sessionId: string): Promise<{ id: string; status: string }> {
    return apiClient.post(`/orchestration/sessions/${sessionId}/cancel`);
  },

  async resumeSession(sessionId: string): Promise<{ id: string; status: string }> {
    return apiClient.post(`/orchestration/sessions/${sessionId}/resume`);
  },

  async listCheckpoints(sessionId: string): Promise<CheckpointRecord[]> {
    return apiClient.get<CheckpointRecord[]>(`/orchestration/sessions/${sessionId}/checkpoints`);
  },

  async requestApproval(taskId: string, reason?: string): Promise<{ id: string; status: string }> {
    const qs = reason ? `?reason=${encodeURIComponent(reason)}` : "";
    return apiClient.post(`/orchestration/tasks/${taskId}/approval${qs}`);
  },

  async resolveApproval(approvalId: string, payload: ApprovalResolvePayload): Promise<{ id: string; status: string }> {
    return apiClient.patch(`/orchestration/approvals/${approvalId}`, payload);
  },
};

File: backend/backend/api/routes/orchestration.py
from fastapi import APIRouter, Depends, HTTPException, Query
from sqlalchemy.ext.asyncio import AsyncSession
from sqlalchemy.future import select
from typing import Optional, List
import json

from backend.database.session import get_db

router = APIRouter()

from backend.schemas.orchestration_schemas import (
    OrchestrationSessionCreate, OrchestrationTaskCreate, ApprovalResolve
)
from backend.services.orchestrator_service import orchestrator_service
from backend.models.orchestration import OrchestrationSession, OrchestrationTask, OrchestrationApproval


# ── Orchestration Endpoints ──────────────────────────────────

@router.post("/orchestration/sessions")
async def create_orchestration_session(payload: OrchestrationSessionCreate, db: AsyncSession = Depends(get_db)):
    session = await orchestrator_service.create_session(db, payload.conversation_id, payload.mode)
    return {
        "id": session.id,
        "conversationId": session.conversation_id,
        "mode": session.mode,
        "status": session.status,
        "sharedContext": session.shared_context,
        "createdAt": session.created_at.isoformat() if session.created_at else None,
    }

@router.get("/orchestration/sessions")
async def list_orchestration_sessions(
    conversation_id: Optional[str] = Query(None),
    status: Optional[str] = Query(None),
    db: AsyncSession = Depends(get_db),
):
    sessions = await orchestrator_service.list_sessions(db, conversation_id, status)
    return [
        {
            "id": s.id,
            "conversationId": s.conversation_id,
            "mode": s.mode,
            "status": s.status,
            "sharedContext": s.shared_context,
            "createdBy": s.created_by,
            "startedAt": s.started_at.isoformat() if s.started_at else None,
            "completedAt": s.completed_at.isoformat() if s.completed_at else None,
            "createdAt": s.created_at.isoformat() if s.created_at else None,
        }
        for s in sessions
    ]

@router.get("/orchestration/sessions/{session_id}")
async def get_orchestration_session(session_id: str, db: AsyncSession = Depends(get_db)):
    session = await orchestrator_service.get_session(db, session_id)
    if not session:
        raise HTTPException(status_code=404, detail="Session not found")
    tasks = await orchestrator_service.get_tasks(db, session_id)
    approvals = await orchestrator_service.get_approvals(db, session_id)
    return {
        "id": session.id,
        "conversationId": session.conversation_id,
        "mode": session.mode,
        "status": session.status,
        "sharedContext": session.shared_context,
        "createdBy": session.created_by,
        "errorMessage": session.error_message,
        "startedAt": session.started_at.isoformat() if session.started_at else None,
        "completedAt": session.completed_at.isoformat() if session.completed_at else None,
        "tasks": [
            {
                "id": t.id,
                "workerRole": t.worker_role,
                "title": t.title,
                "description": t.description,
                "status": t.status,
                "dependsOn": t.depends_on or [],
                "sequenceOrder": t.sequence_order,
                "errorMessage": t.error_message,
                "startedAt": t.started_at.isoformat() if t.started_at else None,
                "completedAt": t.completed_at.isoformat() if t.completed_at else None,
            }
            for t in tasks
        ],
        "approvals": [
            {
                "id": a.id,
                "taskId": a.task_id,
                "status": a.status,
                "reason": a.reason,
                "reviewerNotes": a.reviewer_notes,
                "requestedAt": a.requested_at.isoformat() if a.requested_at else None,
                "resolvedAt": a.resolved_at.isoformat() if a.resolved_at else None,
            }
            for a in approvals
        ],
    }

@router.post("/orchestration/sessions/{session_id}/tasks")
async def add_orchestration_task(session_id: str, payload: OrchestrationTaskCreate, db: AsyncSession = Depends(get_db)):
    session = await orchestrator_service.get_session(db, session_id)
    if not session:
        raise HTTPException(status_code=404, detail="Session not found")
    if session.status not in ("pending", "paused"):
        raise HTTPException(status_code=400, detail=f"Cannot add tasks to session with status '{session.status}'")
    task = await orchestrator_service.add_task(
        db, session_id, payload.worker_role, payload.title, payload.description,
        payload.input_context, payload.depends_on,
    )
    return {
        "id": task.id,
        "workerRole": task.worker_role,
        "title": task.title,
        "status": task.status,
        "sequenceOrder": task.sequence_order,
    }

@router.post("/orchestration/sessions/{session_id}/execute")
async def execute_orchestration_session(session_id: str, db: AsyncSession = Depends(get_db)):
    try:
        session = await orchestrator_service.execute_session(db, session_id)
        return {"id": session.id, "status": session.status}
    except ValueError as e:
        raise HTTPException(status_code=400, detail=str(e))

@router.post("/orchestration/sessions/{session_id}/cancel")
async def cancel_orchestration_session(session_id: str, db: AsyncSession = Depends(get_db)):
    try:
        session = await orchestrator_service.cancel_session(db, session_id)
        return {"id": session.id, "status": session.status}
    except ValueError as e:
        raise HTTPException(status_code=400, detail=str(e))

@router.post("/orchestration/tasks/{task_id}/approval")
async def request_task_approval(task_id: str, reason: str = Query(""), db: AsyncSession = Depends(get_db)):
    approval = await orchestrator_service.request_approval(db, task_id, reason)
    return {"id": approval.id, "status": approval.status}

@router.patch("/orchestration/approvals/{approval_id}")
async def resolve_approval(approval_id: str, payload: ApprovalResolve, db: AsyncSession = Depends(get_db)):
    try:
        approval = await orchestrator_service.resolve_approval(db, approval_id, payload.approved, payload.notes)
        return {"id": approval.id, "status": approval.status}
    except ValueError as e:
        raise HTTPException(status_code=404, detail=str(e))
